watchtower-plugin/src/retrier.rs

```diff
@@ -29,53 +31,61 @@ impl Retrier {
             wt_client,
             max_elapsed_time_secs,
             max_interval_time_secs,
+            pending_appointments: Arc::new(Mutex::new(HashMap::new())),
         }
     }
-    pub async fn manage_retry(&self, mut unreachable_towers: UnboundedReceiver<TowerId>) {
+    pub async fn manage_retry(
+        &self,
+        mut unreachable_towers: UnboundedReceiver<(TowerId, Locator)>,
+    ) {
         log::info!("Starting retry manager");
 
         loop {
-            let tower_id = unreachable_towers.recv().await.unwrap();
+            let (tower_id, locator) = unreachable_towers.recv().await.unwrap();
+            // Not start a retry if the tower is flagged to be abandoned
             {
-                // Not start a retry if the tower has been abandoned
-                let mut wt_client = self.wt_client.lock().unwrap();
+                let wt_client = self.wt_client.lock().unwrap();
+                if !wt_client.towers.contains_key(&tower_id) {
+                    continue;
+                }
                 if wt_client.towers.get(&tower_id).is_none() {
                     log::info!("Skipping retrying abandoned tower {}", tower_id);
                     continue;
                 }
-                wt_client.set_tower_status(tower_id, crate::TowerStatus::TemporaryUnreachable);
             }
 
+            self.add_pending_appointment(tower_id, locator);
+
             log::info!("Retrying tower {}", tower_id);
-            match retry_notify(
+            let r = retry_notify(
                 ExponentialBackoff {
                     max_elapsed_time: Some(Duration::from_secs(self.max_elapsed_time_secs as u64)),
                     max_interval: Duration::from_secs(self.max_interval_time_secs as u64),
                     ..ExponentialBackoff::default()
                 },
-                || async { self.add_appointment(tower_id).await },
+                || async { self.retry_tower(tower_id).await },
                 |err, _| {
                     log::warn!("Retry error happened with {}. {}", tower_id, err);
                 },
             )
-            .await
-            {
+            .await;
+
+            let mut state = self.wt_client.lock().unwrap();
+            self.pending_appointments.lock().unwrap().remove(&tower_id);
```

Comment:
The missed appointment is added in line +57 and all the pending appointments for that tower are cleared in line +74.
This means there is only one appointment in the pending appointments at any given time?
Or maybe I am missing something :/

Reply:
Not, appointments get appended to a collection if the tower is already being retried. I just realized that here something may have been messed up since we were creating a different async task for every `(tower_id, locator)` pair while there should only have been a single task per tower and we should have been appending data to it